File: EASY/Easy_Cpp/path_sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        return dfs(root, 0, targetSum);
    }

    bool dfs(TreeNode* node, int curSum, int targetSum){
        if (!node){
            return false;
        }

        curSum += node->val; // Keep tracking
        if (!node->left && !node->right){ // if leaf node
            return curSum == targetSum;
        }

        return dfs(node->left, curSum, targetSum) ||
                dfs(node->right, curSum, targetSum);

    }

};

File: EASY/path_sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:

        # use sub function
        def dfs(node, curSum):
            # if empty (meaing there does not exist any root-leaf path even if target sum is 0)
            if not node:
                return False
            
            curSum += node.val # keep tracking
            if not node.left and not node.right: # if leaf node
                return curSum == targetSum # true if leaf node is targetsum, false if not
            
            return (dfs(node.left, curSum) or
                    dfs(node.right, curSum)) # recursive

        return dfs(root, 0)